selfdrive/debug/internal/measure_steering_accuracy.py

```diff
@@ -16,66 +17,112 @@ def sigint_handler(signal, frame):
+    elif args.control_type == "indi":
+      control_state = sm['controlsState'].lateralControlState.indiState
+    elif args.control_type == "lqr":
+      control_state = sm['controlsState'].lateralControlState.lqrState
+    elif args.control_type == "angle":
+      control_state = sm['controlsState'].lateralControlState.angleState
+    else:
+      raise ValueError("invalid lateral control type, see help")
 
-    # must be above 10 m/s, engaged and not overriding steering
-    if actual_speed > 10.0 and enabled and not steer_override:
+    v_ego = sm['carState'].vEgo
+    active = sm['controlsState'].active
+    steer = sm['carControl'].actuators.steer
+    standstill = sm['carState'].standstill
+    overriding = sm['carState'].steeringPressed
+    changing_lanes = sm['lateralPlan'].laneChangeState != 0
+    d_path_points = sm['lateralPlan'].dPathPoints
+    # must be engaged, not at standstill, not overriding steering, and not changing lanes
+    if active and not standstill and not overriding and not changing_lanes:
       cnt += 1
 
-      # wait 5 seconds after engage/override
+      # wait 5 seconds after engage / standstill / override / lane change
       if cnt >= 500:
-        # calculate error before rounding
-        actual_angle = sm['controlsState'].angleSteers
-        desired_angle = sm['carControl'].actuators.steeringAngleDeg
-        angle_error = abs(desired_angle - actual_angle)
+        actual_angle = control_state.steeringAngleDeg
+        desired_angle = control_state.steeringAngleDesiredDeg
 
-        # round numbers
+        # calculate error before rounding, then round for stats grouping
+        angle_error = abs(desired_angle - actual_angle)
         actual_angle = round(actual_angle, 1)
         desired_angle = round(desired_angle, 1)
         angle_error = round(angle_error, 2)
         angle_abs = int(abs(round(desired_angle, 0)))
 
-        # collect stats
-        stats[angle_abs]["err"] += angle_error
-        stats[angle_abs]["cnt"] += 1
-        if actual_angle == desired_angle:
-          stats[angle_abs]["="] += 1
-        else:
-          if desired_angle == 0.:
-            overshoot = True
-          else:
-            overshoot = desired_angle < actual_angle if desired_angle > 0. else desired_angle > actual_angle
-          stats[angle_abs]["+" if overshoot else "-"] += 1
+        for group, group_props in all_groups.items():
+          if v_ego > group_props[0]:
+            # collect stats
+            speed_group_stats[group][angle_abs]["cnt"] += 1
+            speed_group_stats[group][angle_abs]["err"] += angle_error
+            speed_group_stats[group][angle_abs]["steer"] += abs(steer)
+            if len(d_path_points):
+              speed_group_stats[group][angle_abs]["dpp"] += abs(d_path_points[0])
+            if control_state.saturated:
+              speed_group_stats[group][angle_abs]["saturated"] += 1
+            if actual_angle == desired_angle:
```

Comment:
if angle_error < 1:
should be more consistent if i didn't skim over anything important. 

1.4 =/= 1.6 but .6 == 1.4

Reply:
or <= .5

Reply:
Don't think I follow. Both `actual_angle` and `desired_angle` are rounded to one decimal point, so the comparison should work as desired. Only `angle_abs` is rounded to int, and that only determines which bucket the stats go in.

Reply:
I misread the precision, (and typo'd my example)
1.04 =/= 1.06 but 1.06 == 1.09.

Reply:
So angle_error <= .05, would be more consistent.

Reply:
I don't think we'll have any `angle_error <= 0.05` because I think (except maybe ZSS-upgraded Toyota?) we have no vehicles with steering angle precisions better than 0.1 degree. It'll either be dead-on equal, or different by >= 0.1 degrees.

Reply:
Ah, fair enough. Although that begs the question, why even have the "==" column. 

Something like <= .5 would provide a more meaningful bin/stat. 

With <=.5, Output like below is much more useful as the over/under are actually measuring a more meaningful value as well.

```
   0° | actuator: 24% | error: 0.51° | -: 12%:- | =: 58%:= | +: 28%:+ | sat cnt:    0 | dppnts: 0.14 | total:   714
   1° | actuator: 19% | error: 0.40° | -: 11%:- | =: 70%:= | +: 17%:+ | sat cnt:    0 | dppnts: 0.09 | total:  1865
   2° | actuator: 18% | error: 0.39° | -: 20%:- | =: 76%:= | +:  3%:+ | sat cnt:    0 | dppnts: 0.08 | total:  2156
   3° | actuator: 30% | error: 0.62° | -: 63%:- | =: 33%:= | +:  3%:+ | sat cnt:    0 | dppnts: 0.08 | total:   737```

Reply:
Desired angle is not limited by precision, so angle error is rarely zero, right?

Reply:
It's rounded here so the match will "work" but 
it's still rarely zero. 

Taking the comparison right at the limit of resolution does mean that rounding error is significant though. Something like 25% error although it should average out.

Reply:
Hehe, yes `angle_error <= 0.05 because I think (except maybe ZSS-upgraded Toyota?)` ZSS goes down to something like 0.0000001 😂